### PR TITLE
[RFC] [DRAFT] 2D GPU API Proposal

### PR DESCRIPTION
**Pull request by danieldegrasse**
## Introduction

This RFC proposes a generic API for interaction with 2D graphics accelerators.
This API is designed to integrate well with a graphics stack like LVGL, but it
can also be used directly by customer applications.

### Problem description

As customer applications continue to demand high performance graphics with
low power consumption, SOC vendors have begun to integrate 2D graphics IP
into their MCU lines. Zephyr currently has no layer to abstract graphics
operations, and leaves performance and power savings on the table without
a framework to enable this IP.

For example, NXP iMX RT MCUs contain a Pixel Processing Pipeline (2D DMA engine)
this engine must be leveraged to obtain acceptable performance when a customer
application rotates their display image. A subset of the functionality available
in this module can be enabled via the DMA API (as seen in `#60064`), but features
like dynamic rotation or blending multiple buffers are not available.

### Proposed change

The 2D accelerator API will provide a common block of operations, which may not
be supported by all graphics IP devices. Graphics stacks like LVGL may
implement support for acceleration using the 2D accelerator API, or customers
may program directly to the 2D accelerator API.

The following key operations should be supported by this API:
- Basic shape drawing
- Basic buffer blending and rotation
- Path based shape drawing
- Path based blending and rotation
- Pixel format conversion

The API is not intended to supersede the display API, but may replace the
`display_write` call in certain applications.

The usage model for an application would look similar to the following:
2d-broad-view

Note that the Graphics stack or user application is expected to fall back to
software rendering (which the 2D accelerator API does not implement) when
the accelerator device does not support a given operation.

### Command Queuing

GPU devices often enable queueing of multiple commands before running a render
to minimize CPU overhead, therefore all API functions are designed to enqueue
commands, which can then be flushed to start a render.

See an example below:
2d-standard-queue

Each render request will set up a command in the hardware's render queue.
`accel2d_render` will then run all queued operations, which will blend `Buffer0`
onto `Buffer1`. `Buffer1` can then be written to the display with
`display_write`. All render request functions are nonblocking.

### Capabilities API

In order to enable a graphics stack to decide which GPU operations it will
attempt to leverage the API for and which it will render using software,
an API will be provided to check GPU capabilities. This will reduce runtime
overhead, as graphics stacks will not call GPU APIs with unsupported
options only to discover software must be used for rendering after some
checks have been run by the driver.

### Basic APIs

For simple 2D DMA engines like the NXP PXP or ST DMA2D, only a limited
set of drawing commands can be supported. Exposing API signatures that these
IP blocks can implement will insure that they can be fully leveraged, and reduce
the runtime checks these drivers would need to ensure they could support a
given API call.

Therefore, a set of basic APIs will be available for GPU IP, including commands
for the following:
- Draw rectangle
- Draw arc
- Draw line
- Blend multiple images

### Path APIs

For more advanced vector graphics engines, APIs will be defined for drawing
arbitrary shapes. These API signatures will allow these IP blocks to be
fully leveraged within graphics stacks.

Path commands are similar to the format SVG uses to describe shapes, like
following:
paths

The following commands will be supported for this use case:
- Draw shape
- Copy image within bounds of shape

### Direct Display Output

Some graphics engines support outputting directly to a connected display,
without the need to stream the final render result back into a framebuffer
in memory. This can be combined with rotation, scaling, or pixel format
conversion. An API will be provided for this purpose, and when used it will
bypass `display_write`

### Multiple Accelerators

For some systems, a specific IP block may be better suited to raster/blit based
graphics tasks, while another IP block may be used for vector operations.
Another use case for this would be using one IP block for display
rotation, while another is enabled for rendering operations, like so:

```
gpu-shared: gpu-node {
	blit-engine = <&gc355>;
	vector-engine = <&gc355>;
	output-engine = <&pxp>; /* Used with gpu2d_write_display */
};
```

The call graph for a GPU "Virtual GPU" implemented using this method might
look like the following when using LVGL:
2d-general-usage

## Detailed API Definition

For a detailed definition of the API proposal, please see the documentation
and header file present with this RFC. The documentation outlines the rationale
behind each 2D accelerator API, as well as which rendering engines would
support it.

## Dependencies

Any implementation of a 2D accelerator API will also need to be integrated with
LVGL in order to provide value for customers using Zephyr, as well as validate
the API is flexible enough to be useful. Zephyr's LVGL module revision
has support for accelerator backends with the merge of `#59656`. Therefore,
the majority of the work required would be integrating the 2D accelerator API
as an LVGL backend

### Alternatives

The primary alternative to supporting GPUs using a standardized API in Zephyr
would be to implement GPUs directly within LVGL. LVGL already has a
defined 2D acceleration backend: https://docs.lvgl.io/8.3/porting/gpu.html.

This backend lacks the ability to perform display rotation automatically,
but it is capable of leveraging the blending and vector graphics features
of the GPUs listed above. Support for automatic display rotation using
hardware is on the roadmap for the next major revision:
https://github.com/lvgl/lvgl/blob/master/docs/ROADMAP.rst#architecture

The primary advantage of this approach is that Zephyr would avoid
reinventing the wheel. There has already been significant work done to
optimize these accelerators within the LVGL backend, and using a GPU API within
Zephyr may only introduce an abstraction layer that reduces performance.

The primary disadvantages of using LVGL as the chosen method are that
Zephyr will be locked into LVGL for GPU support. Additionally, most drivers
will require some level of integration with devicetree and Zephyr OS primitives,
and without a driver class there will be more challenges as to where to
integrate these changes.

### Concerns and Unresolved Questions

The sampling of IP this proposal is based on is rather limited.
Currently, the IP I am aware of and working to support with this RFC includes
the following:

- NXP PXP (2D DMA engine)
  - scaling
  - rotation by 90/180/270 degrees
  - direct output to LCDIF peripheral
  - blending of multiple buffers
  - color space conversions

- ST DMA2D (2D DMA engine)
  - blending of multiple buffers
  - color space conversions

- NXP SmartDMA
  - rotation by 90/180/270 degrees
  - direct output to MIPI-DSI peripheral

- NXP GC355/GCNanoLiteV
  - Path based drawing (Gradient or solid fill)
  - blending of multiple buffers
  - arbitrary scaling, rotation, and translation of buffers

I have also looked at the ST Neochrom VG and Ambiq (Thinksilicon) NEMA Pico
GPU lines. Both appear to use path based drawing, and offer a similar feature
set to the GC355/GCNanoLiteV. However, given lack of available documentation I
cannot confirm the specifics of either of these GPUs.

### CODE REVIEW

File: doc/hardware/peripherals/gpu/gpu2d.rst [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+API Use Cases
+*************
+
+The API is designed to support the following use cases:
+
+* Single GPU capable of implementing some or all of the listed APIs
+* Multiple GPUs, each implementing a subset of the GPU API
+* Optimized software accelerator, capable of implementing the simple drawing
+  and raster APIs
+* GPU engines such as the PXP or SMARTDMA that can write framebuffers directly
```
Comment (yvanderv) [inline doc/hardware/peripherals/gpu/gpu2d.rst:37]: GPU engine is confusing here. Perhaps the term accelerator would be more accurate.

File: doc/hardware/peripherals/gpu/gpu2d.rst [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+any render operations will complete. Buffers configured as outputs for render
+or draw functions should not be modified until the GPU render queue is flushed.
+
+Using Multiple GPUs
+*******************
+
+To use multiple GPUs, we can define a "shared" GPU device in the Zephyr
+devicetree. The definition may look like the following::
+
+    gpu-shared: gpu-node {
```
Comment (yvanderv) [inline doc/hardware/peripherals/gpu/gpu2d.rst:57]: Is this an expected node in the tree? Or can the name be anything and use an alias?

File: doc/hardware/peripherals/gpu/gpu2d.rst [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        blit-engine = <&gc355>;
+        vector-engine = <&gc355>;
+        output-engine = <&pxp>; /* Used with gpu2d_write_display */
```
Comment (yvanderv) [inline doc/hardware/peripherals/gpu/gpu2d.rst:60]: Describing how your subsystem would use the node different scenarios would be helpful. Suppose there's no support for vector graphics on an SoC. The node gpu-shared would not add the property 'vector-engine' but how would the application/overall graphics framework know that SW rendering is needed?

**Comment by danieldegrasse** [comment]
Closing this PR, as an RFC is now open upstream

**Pull request closed without merging**